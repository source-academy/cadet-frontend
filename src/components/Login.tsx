import { Button, ButtonGroup, Card, Icon, NonIdealState, Spinner } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import * as React from 'react';
import { NavLink } from 'react-router-dom';

type LoginProps = DispatchProps & OwnProps;

export type DispatchProps = {
  handleFetchAuth: (ivleToken: string) => void;
  handleLogin: () => void;
};

export type OwnProps = {
  ivleToken?: string;
};

const Login: React.SFC<LoginProps> = props => {
  if (props.ivleToken) {
    startFetchAuth(props.ivleToken, props.handleFetchAuth);
    return (
      <div className="Login pt-dark">
        <Card className="login-card pt-elevation-4">
          <div className="login-body">
            <NonIdealState description="Logging In..." visual={<Spinner large={true} />} />
          </div>
        </Card>
      </div>
    );
  } else {
    return (
      <div className="Login pt-dark">
        <Card className="login-card pt-elevation-4">
          <div className="login-header">
            <h4>
              <Icon icon={IconNames.LOCK} />
              LOGIN
            </h4>
          </div>
          <div className="login-body">
            <ButtonGroup fill={true} vertical={true}>
              {loginButton(props.handleLogin)}
              {playgroundButton}
            </ButtonGroup>
          </div>
        </Card>
      </div>
    );
  }
};

const startFetchAuth = (ivleToken: string, handleFetchAuth: DispatchProps['handleFetchAuth']) =>
  handleFetchAuth(ivleToken);

const loginButton = (handleClick: () => void) => (
  <Button className="pt-large" rightIcon="log-in" onClick={handleClick}>
    Log in with IVLE
  </Button>
);

const playgroundButton = (
  <NavLink to="/playground">
    <Button className="pt-large" rightIcon="code">
      Try out the playground
    </Button>
  </NavLink>
);

export default Login;
