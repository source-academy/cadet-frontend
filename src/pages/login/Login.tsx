import {
  Button,
  ButtonGroup,
  Card,
  Classes,
  H4,
  Icon,
  NonIdealState,
  Spinner,
  SpinnerSize
} from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import classNames from 'classnames';
import * as React from 'react';

export type LoginProps = DispatchProps & OwnProps;

export type DispatchProps = {
  handleFetchAuth: (code: string, providerId?: string) => void;
  handleLogin: (providerId: string) => void;
};

export type OwnProps = {
  providers: Array<{ id: string; name: string }>;
  code?: string;
  providerId?: string;
};

const Login: React.FunctionComponent<LoginProps> = props => {
  const { code, providerId, handleFetchAuth } = props;

  React.useEffect(() => {
    if (code) {
      handleFetchAuth(code, providerId);
    }
  }, [code, providerId, handleFetchAuth]);

  if (code) {
    return (
      <div className={classNames('Login', Classes.DARK)}>
        <Card className={classNames('login-card', Classes.ELEVATION_4)}>
          <div className="login-body">
            <NonIdealState
              description="Logging In..."
              icon={<Spinner size={SpinnerSize.LARGE} />}
            />
          </div>
        </Card>
      </div>
    );
  }

  return (
    <div className={classNames('Login', Classes.DARK)}>
      <Card className={classNames('login-card', Classes.ELEVATION_4)}>
        <div className="login-header">
          <H4>
            <Icon icon={IconNames.LOCK} />
            LOGIN
          </H4>
        </div>
        <div className="login-body">
          <ButtonGroup fill={true} vertical={true}>
            {props.providers.map(({ id, name }) => (
              <LoginButton handleClick={props.handleLogin} name={name} id={id} key={id} />
            ))}
          </ButtonGroup>
        </div>
      </Card>
    </div>
  );
};

const LoginButton = ({
  handleClick,
  id,
  name
}: {
  handleClick: (id: string) => void;
  id: string;
  name: string;
}) => {
  return (
    <Button
      className={Classes.LARGE}
      rightIcon={IconNames.LOG_IN}
      onClick={React.useCallback(() => handleClick(id), [handleClick, id])}
    >
      {`Log in with ${name}`}
    </Button>
  );
};

export default Login;
