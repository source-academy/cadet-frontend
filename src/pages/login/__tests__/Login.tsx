import { shallow } from 'enzyme';

import Login from '../Login';

test('Login renders correctly', () => {
  const props = {
    handleLogin: () => {},
    handleFetchAuth: (code: string, providerId?: string) => {},
    providers: [
      {
        id: 'nusnet_id',
        name: 'LumiNUS'
      }
    ]
  };
  const app = <Login {...props} />;
  const tree = shallow(app);
  expect(tree.debug()).toMatchSnapshot();
});

test('Loading login renders correctly', () => {
  const props = {
    handleLogin: () => {},
    handleFetchAuth: (code: string, providerId?: string) => {},
    code: 'Luminus Code',
    providers: []
  };
  const app = <Login {...props} />;
  const tree = shallow(app);
  expect(tree.debug()).toMatchSnapshot();
});
