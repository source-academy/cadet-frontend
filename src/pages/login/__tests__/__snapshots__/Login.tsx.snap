// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading login renders correctly 1`] = `
"<div className=\\"Login bp3-dark\\">
  <Blueprint3.Card className=\\"login-card bp3-elevation-4\\" elevation={0} interactive={false}>
    <div className=\\"login-body\\">
      <Blueprint3.NonIdealState description=\\"Logging In...\\" icon={{...}} />
    </div>
  </Blueprint3.Card>
</div>"
`;

exports[`Login renders correctly 1`] = `
"<div className=\\"Login bp3-dark\\">
  <Blueprint3.Card className=\\"login-card bp3-elevation-4\\" elevation={0} interactive={false}>
    <div className=\\"login-header\\">
      <Component>
        <Blueprint3.Icon icon=\\"lock\\" />
        LOGIN
      </Component>
    </div>
    <div className=\\"login-body\\">
      <Blueprint3.ButtonGroup fill={true} vertical={true}>
        <LoginButton handleClick={[Function: handleLogin]} name=\\"LumiNUS\\" id=\\"nusnet_id\\" />
      </Blueprint3.ButtonGroup>
    </div>
  </Blueprint3.Card>
</div>"
`;
